### PR TITLE
added solution

### PR DESCRIPTION
**Pull request by floriangeyer21**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+language: java
+jdk:
+  - openjdk11
+script:
+  - mvn clean install
+cache:
+  directories:
+    - $HOME/.m2
```
Comment (margarita-ignatieva) [inline .travis.yml]: empty line

**Comment by margarita-ignatieva** [comment]
Do you check for JDBC Impl for product somewhere?

**Review by margarita-ignatieva — APPROVED** [review]

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    @Override
+    public Product update(Product product) {
+        String query = "UPDATE products SET name = ?, "
+                + "price = ? WHERE product_id = ? AND deleted = FALSE";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setString(1, product.getName());
+            statement.setDouble(2, product.getPrice());
+            statement.setLong(3, product.getId());
```
Comment (margarita-ignatieva) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java:48]: You can avoid duplicating code with a private method

**Review by margarita-ignatieva — APPROVED** [review]

File: src/main/java/com/internet/shop/util/ConnectionUtil.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    static {
+        try {
+            Class.forName("com.mysql.jdbc.Driver");
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException("Can't find MySQL Driver", e);
+        }
+    }
+
+    public static Connection getConnection() {
+        Properties doProperties = new Properties();
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/util/ConnectionUtil.java]:
```suggestion
        Properties dbProperties = new Properties();
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    public List<Product> getAll() {
+        String query = "SELECT * FROM products WHERE deleted = FALSE";
+        List<Product> productList = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            ResultSet resultSet = preparedStatement.executeQuery(query);
+            while (resultSet.next()) {
+                long productId = resultSet.getLong("product_id");
+                Product product = getProductFromResultSet(resultSet);
+                product.setId(productId);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: you've already done it inside of the method getProductFromResultSet
this code is redundant

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, productId);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                Product product = getProductFromResultSet(resultSet);
+                return Optional.of(product);
+            }
+            return Optional.empty();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get product by id", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: let's make message more informative (add some info about product, that wasn't retrieved)

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            statement.setString(1, product.getName());
+            statement.setDouble(2, product.getPrice());
+            statement.setLong(3, product.getId());
+            int successfulUpdate = statement.executeUpdate();
+            if (successfulUpdate != 0) {
+                return product;
+            }
+            throw new SQLException();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can`t update product", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: same here

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setString(1, product.getName());
+            statement.setDouble(2, product.getPrice());
+            statement.setLong(3, product.getId());
+            int successfulUpdate = statement.executeUpdate();
+            if (successfulUpdate != 0) {
+                return product;
+            }
+            throw new SQLException();
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: let's throw out custom exception here + add some message

**Review by RomanDubovskyi — COMMENTED** [review]
add please main method to pr to show, where you test your code

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            statement.setString(1, product.getName());
+            statement.setDouble(2, product.getPrice());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                newProduct.setId(resultSet.getLong(1));
+            }
+            return newProduct;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't create product", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Can't create product with name " + product.getName() , e);
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    @Override
+    public Product update(Product product) {
+        String query = "UPDATE products SET name = ?, "
+                + "price = ? WHERE product_id = ? AND deleted = FALSE";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setString(1, product.getName());
+            statement.setDouble(2, product.getPrice());
+            statement.setLong(3, product.getId());
+            int successfulUpdate = statement.executeUpdate();
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            int changedRows = statement.executeUpdate();
```

File: src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        String query = "SELECT * FROM products WHERE deleted = FALSE";
+        List<Product> productList = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement preparedStatement = connection.prepareStatement(query);
+            ResultSet resultSet = preparedStatement.executeQuery(query);
+            while (resultSet.next()) {
+                productList.add(getProductFromResultSet(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Incorrect getAll query", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Can't get all products ", e);
```

**Review by RomanDubovskyi — APPROVED** [review]

**Pull request merged**